tree/treeplayer/inc/TTreeReaderArray.h

```diff
@@ -59,90 +59,159 @@ namespace Internal {
+      Iterator_t() : fArray(nullptr), fIndex(0u), fSize(0u) {}
+
+      /// Construct iterator
+      Iterator_t(std::size_t index, TTreeReaderArray *array)
+         : fArray(array), fIndex(index), fSize(fArray ? fArray->GetSize() : 0u)
+      {
+         if (fIndex >= fSize)
+            fArray = nullptr; // invalidate iterator
+      }
 
-   // Iterator through the indices of a TTreeReaderArray.
-   struct Iterator_t:
-      public std::iterator<std::input_iterator_tag, T, long> {
-      // Default initialized, past-end iterator.
-      Iterator_t() :
-         fIndex(0), fArray(0) {}
+      /// Construct iterator from a const TTreeReaderArray
+      Iterator_t(std::size_t index, const TTreeReaderArray *array)
+         : Iterator_t(index, const_cast<TTreeReaderArray *>(array)) {}
 
-      // Initialize with an array and index.
-      Iterator_t(size_t idx, TTreeReaderArray* array) :
-         fIndex(idx), fArray(array) {}
+      Iterator_t(Iterator_t &&) = default;
+      Iterator_t(const Iterator_t &) = default;
+      Iterator_t &operator=(Iterator_t &&) = default;
+      Iterator_t &operator=(const Iterator_t &) = default;
 
-      size_t fIndex; // Current index in the array.
-      TTreeReaderArray* fArray; // The array iterated over; 0 if invalid / end.
+      T &operator*()
+      {
+         R__ASSERT(fArray && "invalid iterator!");
+         return fArray->At(fIndex);
+      }
 
-      bool IsValid() const { return fArray; }
+      const T &operator*() const
+      {
+         R__ASSERT(fArray && "invalid iterator!");
+         return fArray->At(fIndex);
+      }
+
+      T *operator->() { return IsValid() ? &fArray->At(fIndex) : nullptr; }
 
-      bool operator==(const Iterator_t& lhs) const {
-         // Compare two iterators as equal; follow C++14 requiring two past-end
-         // iterators to be equal.
-         if (!IsValid() && !lhs.IsValid())
+      bool operator==(const Iterator_t &other) const
+      {
+         // Follow C++14 requiring two past-the-end iterators to be equal.
+         if (!IsValid() && !other.IsValid())
             return true;
-         return fIndex == lhs.fIndex && fArray == lhs.fArray;
+         return fArray == other.fArray && fIndex == other.fIndex;
       }
 
-      bool operator!=(const Iterator_t& lhs) const {
-         // Compare not equal.
-         return !(*this == lhs);
+      bool operator!=(const Iterator_t &other) const { return !(*this == other); }
+
+      /// Pre-increment operator
+      Iterator_t &operator++()
+      {
+         ++fIndex;
+         if (fIndex >= fSize)
+            fArray = nullptr; // invalidate iterator
+         return *this;
       }
 
-      Iterator_t operator++(int) {
-         // Post-increment (it++).
-         Iterator_t ret = *this;
-         this->operator++();
+      /// Post-increment operator
+      Iterator_t operator++(int)
+      {
+         auto ret = *this;
+         auto &self = *this;
+         ++self;
```

Comment:
You find these two lines nicer than `this->operator++()` or if need be `++(*this)`? I find the option you chose my third favorite, due to the `auto &` and the second local var that seems to be almost identical to `ret`. I.e. I find this code much more complex to grasp due to the extra degree of freedom (extra local var)... But as "I find" tries to say: this is really a matter of taste, so if you disagree then feel free to keep yours!

Reply:
I agree, changed